File: cli/src/commands/scan.js
import path from 'path';
import { scanProject } from '../scanner.js';
import { saveGraph, computeFileHash } from '../graph.js';
import { saveSlices } from '../slicer.js';

/**
 * Try to get the current git commit hash from the given directory.
 * Returns null if the directory is not a git repo or simple-git is unavailable.
 *
 * @param {string} dir - Directory to check for git.
 * @returns {Promise<string|null>} The current commit hash or null.
 */
async function getGitCommitHash(dir) {
  try {
    const { simpleGit } = await import('simple-git');
    const git = simpleGit(dir);
    const isRepo = await git.checkIsRepo();
    if (!isRepo) return null;
    const log = await git.log({ maxCount: 1 });
    return log.latest ? log.latest.hash : null;
  } catch {
    return null;
  }
}

/**
 * Register the `scan` command on the given Commander program.
 *
 * @param {import('commander').Command} program - The Commander program instance.
 */
export function registerScanCommand(program) {
  program
    .command('scan [dir]')
    .description('Scan a project directory and generate a code graph')
    .option('--exclude <patterns...>', 'Additional glob patterns to exclude')
    .action(async (dir, options) => {
      const rootDir = path.resolve(dir || '.');
      const outputDir = path.join(rootDir, '.codemap');

      const startTime = Date.now();

      // Step 1: Scan the project
      const graph = await scanProject(rootDir, {
        exclude: options.exclude || [],
      });

      // Step 2: Get git commit hash (gracefully skip)
      const commitHash = await getGitCommitHash(rootDir);
      graph.commitHash = commitHash;

      const scanDuration = Date.now() - startTime;

      // Step 3: Build file hashes map
      const fileHashes = {};
      for (const [filePath, fileData] of Object.entries(graph.files)) {
        fileHashes[filePath] = fileData.hash;
      }

      // Step 4: Build meta object
      const meta = {
        lastScanAt: new Date().toISOString(),
        commitHash,
        scanDuration,
        fileHashes,
      };

      // Step 5: Save graph and meta
      await saveGraph(outputDir, graph, meta);

      // Step 6: Save slices
      await saveSlices(outputDir, graph);

      // Step 7: Print summary
      const fileCount = graph.summary.totalFiles;
      const funcCount = graph.summary.totalFunctions;
      const moduleNames = graph.summary.modules.join(', ');

      console.log(`Scan complete.`);
      console.log(`  Files:     ${fileCount}`);
      console.log(`  Functions: ${funcCount}`);
      console.log(`  Modules:   ${moduleNames}`);
      console.log(`  Output:    ${outputDir}`);
    });
}

File: rust-cli/src/commands/scan.rs
use clap::Args;
use std::path::PathBuf;

#[derive(Args)]
pub struct ScanArgs {
    /// Project directory to scan
    pub dir: Option<String>,
    /// Additional glob patterns to exclude
    #[arg(long, num_args = 1..)]
    pub exclude: Vec<String>,
}

pub fn run(args: ScanArgs) {
    let dir = args.dir.unwrap_or_else(|| ".".to_string());
    let root = PathBuf::from(&dir);
    let root = match root.canonicalize() {
        Ok(p) => p,
        Err(e) => {
            eprintln!("Error: cannot resolve directory '{}': {}", dir, e);
            std::process::exit(1);
        }
    };

    println!("Scanning {}...", root.display());

    match crate::scanner::scan_and_save(&root, &args.exclude) {
        Ok(graph) => {
            let codemap_dir = root.join(".codemap");
            // 生成 slices/（与 Node.js scan 行为一致）
            if let Err(e) = crate::slicer::save_slices(&codemap_dir, &graph) {
                eprintln!("Warning: failed to save slices: {}", e);
            }
            println!("Scan complete.");
            println!("  Files:     {}", graph.summary.total_files);
            println!("  Functions: {}", graph.summary.total_functions);
            println!("  Modules:   {}", graph.summary.modules.join(", "));
            println!("  Output:    {}", codemap_dir.display());
        }
        Err(e) => {
            eprintln!("Scan failed: {}", e);
            std::process::exit(1);
        }
    }
}
